### PR TITLE
Force Servo Process

### PR DESCRIPTION
**Pull request by aditn**
This is a large pull request that moves the force servoing process from python to GRBL. Now the firmware can accept a command followed by a target force value (via serial port) and the gripper motor will move in one motion until that target force value is achieved. This will reduce the time it takes to bump keys significantly and will reduce the messages received by the brain.

EDIT:
Order for Viewing:
- `system.c`: This is where the force servoing command is defined. This finds the command, grabs the
  float directly after (which is the target voltage), and executes the force servoing procedure.
- `motion_control.c`: This is the wrapper function for the force servoing procedure and mimics the
  wrapper function for the homing procedure.
- `limits.c`: This is where the motion is sent to a planning block followed by waking up the motors.
- `stepper.c`: The additions to the stepper motor ISR are for stopping the gripper when the desired
  force value is reached.
- `report.c`: This is where the function for retrieving force and motor voltage values are. Now
  `calculate_force_voltage()` is constantly called in `protocol.c`, while `calculate_motor_voltage` is called only when
  a voltage request is made.

### CODE REVIEW

File: limits.c [hunk 1]
```diff
@@ -160,7 +163,7 @@ void limits_go_home(uint8_t cycle_mask)
         target[idx] = 0;
       }
     }
-
+
```
Comment (penguinland) [inline limits.c]: Get rid of this space.

**Comment by penguinland** [comment]
Next time, try to make smaller PRs. One 500-line PR is much harder to review than five 100-line PRs. Do you have suggestions for the order we should read these files? or other tips for reviewing?

File: motion_control.c [hunk 2]
```diff
@@ -257,6 +257,29 @@ void mc_homing_cycle(uint8_t axis_mask)
...
+  // Force Servoing Complete! Setup system for normal operation.
+
+  // Gcode parser position was circumvented by the limits_force_servo() routine, so sync position now.
+  gc_sync_position();
+  
+  // Set idle state after servoing completes and before returning to main program.  
+  sys.state = STATE_IDLE;
+  st_go_idle();
+
+  // If hard limits feature enabled, re-enable hard limits pin change register after cycle.
```
Comment (penguinland) [inline motion_control.c]: I like all the comments in here. Thanks for explaining what's happening!

File: protocol.c [hunk 3]
```diff
@@ -116,48 +116,49 @@ void protocol_main_loop()
...
-          // Throw away all comment characters
-          if (c == ')') {
-            // End of comment. Resume line.
-            iscomment = false;
+          if (iscomment) {
+            // Throw away all comment characters
+            if (c == ')') {
+              // End of comment. Resume line.
+              iscomment = false;
+            }
```
Comment (penguinland) [inline protocol.c]: This looks like you just changed the indentation of the code without changing anything significant. Is that correct? If so, undo it; large changes to the whitespace formatting should go in a separate PR that doesn't also have changes to functionality in it. 

This comment applies to lines 119-158ish.

File: protocol.c [hunk 4]
```diff
@@ -271,8 +272,8 @@ void protocol_execute_runtime()
     }
 
     // Execute a cycle start by starting the stepper interrupt begin executing the blocks in queue.
-    // block Start while homing.
-    if ((rt_exec & EXEC_CYCLE_START) && !(sys.state & STATE_HOMING)) {
+    // block Start while homing/force-servoing.
+    if ((rt_exec & EXEC_CYCLE_START) && !(sys.state & STATE_HOMING)&& !(sys.state & STATE_FORCESERVO)) {
```
Comment (penguinland) [inline protocol.c]: Space before the `&&` so it's consistent with the rest of the line.

File: report.c [hunk 5]
```diff
@@ -109,6 +109,7 @@ void report_alarm_message(int8_t alarm_code)
   if (alarm_code & ALARM_PROBE_FAIL)  printPgmString(PSTR("Probe fail "));
   if (alarm_code & ALARM_HOME_FAIL)   printPgmString(PSTR("Homing fail "));
   if (alarm_code & ALARM_ESTOP)       printPgmString(PSTR("Estop pressed "));
+  if (alarm_code & ALARM_FORCESERVO_FAIL) printPgmString(PSTR("Force servoing fail"));
```
Comment (penguinland) [inline report.c:112]: Add spaces to the previous lines so all the `printPgmString`s line up again.

File: settings.h [hunk 6]
```diff
@@ -26,7 +26,7 @@
 
 
 // define GRBL version and build
-#define GRBL_VERSION "0.9K.3.3"
+#define GRBL_VERSION "0.9K.3.Force"
```
Comment (penguinland) [inline settings.h]: @mdicicco, is it okay if the version number isn't a number like this?

File: system.c [hunk 7]
```diff
@@ -49,6 +50,25 @@ void system_init()
...
+  TCCR1A &= ~((1<<WGM11) | (1<<WGM10));
+  TCCR1A &= ~((1<<COM1A1) | (1<<COM1A0) | (1<<COM1B1) | (1<<COM1B0)); // Disconnect OC4 output
+
+  // Timer Match value of 0xFFFF. This can be reduced to retrieve voltage values sooner
+  OCR1A = 0XF0;
+  OCR1B = 0X00;
+
+  // Initialize ADC for Voltage Monitoring
+  init_ADC();*/
+  /* KEYME SPECIFIC END */
```
Comment (penguinland) [inline system.c]: You have added a whole bunch of commented-out code here. Either remove it or uncomment it (or explain why this isn't currently used but will be uncommented in the near future)

File: system.c [hunk 8]
```diff
@@ -57,6 +77,56 @@ ISR(TIMER2_COMPA_vect)
...
+    ADCSRB &= ~(1<<MUX5_BIT); // Clear MUX5_BIT which is set when force sensor is target
+    ADMUX = (1<<REFS0) + voltage_result_index; // set next motor target for ADC
+  }
+  else{
+    // set force sensor as next target
+    ADMUX = (1<<REFS0) + FVOLT_ADC;
+    ADCSRB |= (1<<MUX5_BIT);
+  }
+}*/
+/* KEY ME SPECIFIC END*/
```
Comment (penguinland) [inline system.c]: Same comment here.

File: system.h [hunk 9]
```diff
@@ -161,5 +165,7 @@ enum {
 #define TIME_ON(tid)  (((tid)==ACTIVE_TIMER)?(TIMING_PORT&=~TIMING_MASK):0)
 #define TIME_TOGGLE(tid)  (((tid)==ACTIVE_TIMER)?(TIMING_PIN|=TIMING_MASK):0)
 
-
+// Voltage Monitoring
+#define VOLTAGE_SENSOR_COUNT 5 // number of sensors we are probing
+//void init_ADC();
```
Comment (penguinland) [inline system.h:171]: As a general rule, don't include commented-out code in pull requests.

File: system.h [hunk 10]
```diff
@@ -134,13 +137,14 @@ void system_execute_startup(char *line);
 
 // But We don't need such a big value, and we do need the high bit free
 //#define MAX_LINE_NUMBER 9999999
-#define LINENUMBER_EMPTY_BLOCK 0x8000 //the other bit, used as a flag
-#define LINENUMBER_SPECIAL      0x4000
+#define LINENUMBER_EMPTY_BLOCK 0x10000 //the other bit, used as a flag
+#define LINENUMBER_SPECIAL      0x4000 //denotes Homing and Probing among a couple other processes
+#define LINENUMBER_SPECIAL_SERVO 0x8000 //denotes force servoing
```
Comment (penguinland) [inline system.h]: I'm a little uncomfortable redefining constants like this. Why not make the force servoing bit 0x10000 and leave 0x8000 as the empty block?

edit: the reason I'm a little uncomfortable is that there might be code making implicit assumptions about the value of `LINENUMBER_EMPTY_BLOCK` that could break if you change it. I am not aware of this code, but there's a lot of GRBL I haven't looked at.

File: settings.h (hunk 6 above)
Comment (mdicicco) [inline settings.h]: no, that's a test feature that he will need to remove before merging

File: system.h (hunk 10 above)
Comment (mdicicco) [inline system.h]: We're aware that this is a concern.  Adit and I are going to address this once we get the whole PR cleaned up and ready.

File: .gitignore [hunk 11]
```diff
@@ -1,3 +1,4 @@
+testSerial.py
```
Comment (kborer) [inline .gitignore]: why?

File: limits.c [hunk 12]
```diff
@@ -243,6 +246,7 @@ void limits_go_home(uint8_t cycle_mask)
   plan_buffer_line(target, min_seek_rate, false, LINENUMBER_SPECIAL|(homing_line_number*4+LINEMASK_DONE));
   homing_line_number++;
 
+
```
Comment (kborer) [inline limits.c]: rm

File: limits.c [hunk 13]
```diff
@@ -254,7 +258,6 @@ void limits_go_home(uint8_t cycle_mask)
   sys.state = STATE_HOMING;
 }
 
-
```
Comment (kborer) [inline limits.c:257]: put back

File: .gitignore (hunk 11 above)
Comment (mdicicco) [inline .gitignore]: I think Adit may just be unclear about what the git ignore file is for.  I'll clarify for him.  He probably wrote this as a test file and didn't want to commit it.

File: limits.c [hunk 14]
```diff
@@ -283,3 +286,78 @@ void limits_soft_check(float *target)
...
+// for bumping. Future iterations may want to include a specific force to reach in order to
+// do any force related movements. More motor motions may be added to reach the desired force
+// more accurately.
+void limits_force_servo(){
+  if (sys.abort) { return; } // Block if system reset has been issued.
+
+  uint16_t bump_target_force = force_target_val; // This is the input target force sensor value
+  float servo_rate;
+  // Initialize homing in search mode to quickly engage the specified cycle_mask limit switches.
+  uint8_t approach=~0;
```
Comment (kborer) [inline limits.c]: use spaces around binary operators

Comment (kborer) [inline limits.c]: e.g. 

``` c
uint8_t approach = ~0;
```

File: limits.c [hunk 15]
```diff
@@ -283,3 +286,78 @@ void limits_soft_check(float *target)
...
+
+  // axislock bit is high if axis is homing, so we only enable checking on moving axes.
+  limits_enable(axislock,~approach);  //expect 0 on approach (stop when 1). vice versa for pulloff
+  limits.isservoing = 1; // tell system we are servoing
+
+  st_prep_buffer(); // Prep and fill segment buffer from newly planned block.
+  st_wake_up(); // Initiate motion
+
+  do {
+    //calculate_voltage();
```
Comment (kborer) [inline limits.c]: rm

File: limits.c [hunk 16]
```diff
@@ -283,3 +286,78 @@ void limits_soft_check(float *target)
...
+    protocol_execute_runtime();
+    if (SYS_EXEC & EXEC_RESET) {
+      protocol_execute_runtime();
+      return;
+    }
+
+    // Check if we never reached limit switch.  call it a Probe fail.
+    if (SYS_EXEC & EXEC_CYCLE_STOP) {
+      sys.alarm |= ALARM_FORCESERVO_FAIL;
+      SYS_EXEC|=EXEC_CRIT_EVENT;
```
Comment (kborer) [inline limits.c]: spacing

File: protocol.h [hunk 17]
```diff
@@ -31,6 +31,30 @@
   #define LINE_BUFFER_SIZE 70
 #endif
 
+
+// While enabled, force servoing is in process
+extern volatile uint8_t force_servo_enable;
+// While enabled, recieving values to store in fservodata struct
```
Comment (kborer) [inline protocol.h]: receiving

File: serial.c [hunk 18]
```diff
@@ -40,7 +40,6 @@ volatile uint8_t tx_buffer_tail = 0;
 
 uint8_t checksum = 0;  //sum all bytes between newlines.
 
-
```
Comment (kborer) [inline serial.c:43]: put this back

File: stepper.c [hunk 19]
```diff
@@ -332,13 +332,15 @@ ISR(TIMER1_COMPA_vect)
...
-  // exactly settings.pulse_microseconds microseconds, independent of the main Timer1 prescaler.
+  // exactly settings.pulse_microseconds microseconds, independent of the main Timer4 prescaler.
   TCNT0 = st.step_pulse_time; // Reload Timer0 counter
   TCCR0B = (1<<CS01); // Begin Timer0. Full speed, 1/8 prescaler
 
   busy = true;
   sei(); // Re-enable interrupts to allow Stepper Port Reset Interrupt to fire on-time.
          // NOTE: The remaining code in this ISR will finish before returning to main program.
+  
+  //disable_ADC_ISR(); // Disable ADC ISR to avoid firing before Stepper Port Reset ISR
```
Comment (kborer) [inline stepper.c]: rm

File: stepper.c [hunk 20]
```diff
@@ -672,14 +697,14 @@ void stepper_init()
...
   DIRECTION_PORT = (DIRECTION_PORT & ~DIRECTION_MASK) | settings.dir_invert_mask;
-
-  // Configure Timer 1: Stepper Driver Interrupt
-  TCCR1B &= ~(1<<WGM13); // waveform generation = 0100 = CTC
-  TCCR1B |=  (1<<WGM12);
-  TCCR1A &= ~((1<<WGM11) | (1<<WGM10));
-  TCCR1A &= ~((1<<COM1A1) | (1<<COM1A0) | (1<<COM1B1) | (1<<COM1B0)); // Disconnect OC1 output
-  // TCCR1B = (TCCR1B & ~((1<<CS12) | (1<<CS11))) | (1<<CS10); // Set in st_go_idle().
-  // TIMSK1 &= ~(1<<OCIE1A);  // Set in st_go_idle().
+
```
Comment (kborer) [inline stepper.c]: rm whitespace

File: limits.c [hunk 21]
```diff
@@ -283,3 +286,77 @@ void limits_soft_check(float *target)
...
+  }
+  else if(voltage_result[4]>(bump_target_force+GRIPPER_FORCE_THRESHOLD)){
+    travel_servo = -MAXSERVODIST;
+  }
+  //approach has all bits set (negative dir) or none (positive)
+  float target[N_AXIS];
+  
+  // Determine travel distance to the furthest homing switch based on user max travel settings.
+  uint8_t axislock = AXISLOCKSERVO;
+  servo_rate = settings.homing_seek_rate[X_AXIS]/12.0; // TODO: make this into macro for servoing
```
Comment (mdicicco) [inline limits.c]: What's the 12.0 here do?

File: limits.h [hunk 22]
```diff
@@ -26,11 +26,14 @@
 // This is defined by the largest homing flag in the system. The Y flag is
 // the largest with a length of just under 40 mm.
 #define MAXFLAGLEN 40.0
+#define MAXSERVODIST 15.0
```
Comment (mdicicco) [inline limits.h]: We should probably discuss if this is smart.  Full travel is about 11 or 12.  Do we want to be able to do this move from zero?

File: report.c [hunk 23]
```diff
@@ -378,48 +380,47 @@ void report_counters()
...
 }
 
-//Prints voltage data: motor volts.
+// Prints voltage data for Motors (C,X,Y,Z) and Force sensor.
 void report_voltage()
 {
+  uint8_t i;
+  calculate_motor_voltage();
+  printPgmString(PSTR("|"));
+  for (i = 0; i<VOLTAGE_SENSOR_COUNT; i++){
```
Comment (mdicicco) [inline report.c:389]: I am nervous about reading all 5 sensors on every tick.  There seems to be no reason to do this.  I think it's better to read one at a time, per tick.

Either: 
x->y->z->c->F->repeat
or
x->F->y->F->z->F->c->F->repeat

Comment (mdicicco) [inline report.c:389]: actually, nevermind... this is in report... I have to think about this more

File: system.h [hunk 24]
```diff
@@ -134,13 +137,14 @@ void system_execute_startup(char *line);
...
 #define LINENUMBER_MAX         (LINENUMBER_SPECIAL-1)
 #define LINEMASK_OFF_EDGE         (0x0)
 #define LINEMASK_ON_EDGE         (0x1)
 #define LINEMASK_DONE           (0x2)
-typedef uint16_t linenumber_t;  //resize back to int32 for bigger numbers
+typedef uint32_t linenumber_t;  //resize back to int32 for bigger numbers
```
Comment (mdicicco) [inline system.h]: can probably remove this old comment now

File: system.h [hunk 25]
```diff
@@ -161,5 +165,8 @@ enum {
 #define TIME_ON(tid)  (((tid)==ACTIVE_TIMER)?(TIMING_PORT&=~TIMING_MASK):0)
 #define TIME_TOGGLE(tid)  (((tid)==ACTIVE_TIMER)?(TIMING_PIN|=TIMING_MASK):0)
 
-
+// Voltage Monitoring
+#define VOLTAGE_SENSOR_COUNT 5 // number of sensors we are probing
```
Comment (mdicicco) [inline system.h]: comment here isn't quite right.  This is the number of things we're measuring with ADC input pins.  The word 'probe' is reserved for something else in this context

File: system.h (hunk 10 above)
Comment (mdicicco) [inline system.h]: @aditn and I talked about this today.  It's better to switch these back so that empty is 0x8000 again and servo ix 0x10000.  Adit is going to test that.  If it works, that is preferred so that we can maintain backwards compatibility and 3.4 can run without any python code changes.

File: report.h [hunk 26]
```diff
@@ -92,6 +92,12 @@ void report_counters();
 
 // Prinst voltage of motors
 void report_voltage();
+void calculate_motor_voltage();
+void calculate_force_voltage();
+// Variable for holding voltage value after ADC
+extern uint16_t voltage_result[VOLTAGE_SENSOR_COUNT];
```
Comment (mdicicco) [inline report.h]: You might want to give this a better name.  Maybe analog_voltage_readings, analog_voltages, analog_voltage_values, andalog_read_storage

This is global storage for analogs so I would like the variable name to reflect that immediately in the variable name.

File: report.c [hunk 27]
```diff
@@ -378,48 +380,47 @@ void report_counters()
   printPgmString(PSTR("}\r\n"));
 }
 
-//Prints voltage data: motor volts.
+// Prints voltage data for Motors (C,X,Y,Z) and Force sensor.
```
Comment (mdicicco) [inline report.c]: let the comment say 'motor bus voltage'

File: report.c [hunk 28]
```diff
@@ -432,26 +433,19 @@ void report_voltage()
   ADCSRA |= (1<<ADSC);
   // Wait for the result
   while(ADCSRA & (1<<ADSC));
-  // Print the result
-  printInteger(ADC);
-
+  // Store the result
+  voltage_result[4] = ADC;
```
Comment (mdicicco) [inline report.c]: might be a good idea to make this "4" be a defined const for clarity

File: limits.c [hunk 29]
```diff
@@ -310,7 +310,7 @@ void limits_force_servo(){
   float target[N_AXIS];
   
   uint8_t axislock = AXISLOCKSERVO;
-  servo_rate = settings.homing_seek_rate[X_AXIS]/16.0; // TODO: make servo_rate into macro for servoing
+  servo_rate = settings.homing_seek_rate[X_AXIS]/20.0; // TODO: make servo_rate into macro for servoing
```
Comment (mdicicco) [inline limits.c:313]: What's the 20.0 here?

**Comment by mdicicco** [comment]
I finished going through this.  Some minor formatting issues, but I don't want to get distracted by that.  Overall I think we're good.  Just address the outstanding issues and squash and I am okay to merge.

File: report.c [hunk 30]
```diff
@@ -406,14 +406,14 @@ void calculate_motor_voltage(){
 
   for (i=0;i<(VOLTAGE_SENSOR_COUNT-1); i++){
     ADCSRB &= ~(1<<MUX5_BIT); // Clear MUX5_BIT which is set when force sensor is target
-    ADMUX = (1<<REFS0) + voltage_result_index; // set next motor target for ADC
```
Comment (mdicicco) [inline report.c]: heh, this looks like an important change

File: system.c [hunk 31]
```diff
@@ -31,8 +31,8 @@
 
 
 uint32_t masterclock=0;
-uint16_t voltage_result[VOLTAGE_SENSOR_COUNT];
-uint8_t voltage_result_index = 0;
+//uint16_t voltage_result[VOLTAGE_SENSOR_COUNT];
```
Comment (mdicicco) [inline system.c:34]: just delete these

File: report.c (hunk 30 above)
Comment (aditn) [inline report.c]: `voltage_result_index` was used when I had the ADC on an ISR. But in this situation `voltage_result_index` doesn't get incremented while `i` does. I should be using `i` in this situation otherwise I will be printing the voltages of X,Y,Y,Y,F instead of X,Y,Z,C,F.
